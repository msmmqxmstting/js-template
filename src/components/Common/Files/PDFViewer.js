/**
 * Created by xun on  2021/9/1 16:07.
 * description: PDFViewer
 */
import React, { useEffect, useRef } from 'react';
export class PDFJSBackend {
	init = (source, element, iframeSrcPrefix = '') => {
		const iframe = document.createElement('iframe');
		iframe.id = 'hxPDFIframe';
		iframe.src = `${iframeSrcPrefix}/hx-pdfjs-dist/web/viewer.html?file=${source}`;
		iframe.width = '100%';
		iframe.height = '100%';
		if (element && element.length > 0) {
			return;
		}
		element.appendChild(iframe);
	};
}

const PDFViewer = ({
	isFromModal = false,
	iframeSrcPrefix,
	pdfFile,
	containerStyle = { height: 600 }
}) => {
	if (!pdfFile) {
		return '';
	}
	const ref = useRef(null);
	const backend = new PDFJSBackend();
	const setNodesAttrs = () => {
		const timeoutFunc = () => {
			const iframe = window.document.getElementById('hxPDFIframe');
			const innerDoc = iframe.contentDocument || iframe.contentWindow.document;
			if (!innerDoc) {
				return;
			}
			const openFileBtn = innerDoc.getElementById('openFile');

			if (!openFileBtn) {
				return;
			}
			openFileBtn.style.display = 'none';
		};

		const tid = setTimeout(timeoutFunc, 1000);
		clearTimeout(tid);
	};

	useEffect(() => {
		backend.init(pdfFile, ref.current, iframeSrcPrefix);
		setNodesAttrs();
		return () => {
			ref.current.innerHTML = '';
		};
	}, [pdfFile]);

	return (
		<div>
			<div
				ref={ref}
				id="hx-pdf-viewer"
				style={{
					width: '100%',
					...containerStyle,
					height: isFromModal ? '100vh' : containerStyle.height || '100%'
				}}
			/>
		</div>
	);
};
export default PDFViewer;
