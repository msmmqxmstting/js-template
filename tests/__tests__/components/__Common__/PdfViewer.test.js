/**
 * Created By xun  on 2018-11-20 08:54.
 * Description: __Login__.test
 */

import React from 'react';

import { shallowWrap } from '../../../tools/contextWrap';

import PDFViewer from 'components/Common/Files/PDFViewer';

describe('PDF预览组件', () => {
	let useEffect;

	const mockUseEffect = () => {
		useEffect.mockImplementationOnce((f) => f());
	};
	it('PDFViewer正确渲染快照', () => {
		const wrapper = shallowWrap(<PDFViewer pdfFile={'pdf file url'} />);
		expect(wrapper.html()).toMatchSnapshot();
	});
	it('PDFViewer缺少pdfFile属性', () => {
		const wrapper = shallowWrap(<PDFViewer />);
		expect(wrapper.text()).toBe('');
	});
/*	it('pdfFile属性改变触发useEffect', () => {
		const wrapper = shallowWrap(<PDFViewer pdfFile={'pdf file url'} />);
		wrapper.setProps({ pdfFile: 'new pdf file url' });
		useEffect = jest.spyOn(React, 'useEffect');
		mockUseEffect();
		const instance = wrapper.instance();
		const spyFn = jest.spyOn(instance, 'setNodesAttrs');
		instance.setNodesAttrs();
		expect(spyFn).toHaveBeenCalled();
	});*/
});
