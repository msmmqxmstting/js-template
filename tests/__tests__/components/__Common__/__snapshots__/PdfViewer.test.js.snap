// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PDF预览组件 PDFViewer正确渲染快照 1`] = `"<div><div id=\\"hx-pdf-viewer\\" style=\\"width:100%;height:600px\\"></div></div>"`;
